### PR TITLE
Improved Recipient details screen -> pubic key details screen

### PR DESCRIPTION
**Pull request by DenBond7**
This PR improved the pubic key details screen

issue #2469 

image

Before changes

image

After changes

image

useful links

1. https://davesteele.github.io/gpg/2014/09/20/anatomy-of-a-gpg-key/#OtherTools

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities

### CODE REVIEW

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/bouncycastle/openpgp/PGPPublicKeyExt.kt [hunk 1]
```diff
@@ -20,3 +27,62 @@ fun PGPPublicKey.armor(
...
+  }
+
+  for (signature in keySignatures) {
+    if (mostRecent == null || signature.creationTime.after(mostRecent.creationTime)) {
+      mostRecent = signature
+    }
+  }
+
+  return mostRecent?.creationTime ?: creationTime
+}
```
Comment (ioanmo226) [inline FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/bouncycastle/openpgp/PGPPublicKeyExt.kt:46]: We may simplify like below.

```kt
fun PGPPublicKey.getLastModificationDate(): Date {
  val allSignatures = (listOf(signatures, keySignatures).flatMap { it.asSequence() }).toList()
  return allSignatures.maxByOrNull { it.creationTime }?.creationTime ?: creationTime
}
```

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/bouncycastle/openpgp/PGPPublicKeyExt.kt [hunk 2]
```diff
@@ -20,3 +27,62 @@ fun PGPPublicKey.armor(
...
+
+fun PGPPublicKey.getExpirationDate(): Date? {
+  return if (validSeconds == 0L) {
+    null
+  } else {
+    Date(creationTime.time + TimeUnit.SECONDS.toMillis(validSeconds))
+  }
+}
```
Comment (ioanmo226) [inline FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/bouncycastle/openpgp/PGPPublicKeyExt.kt:54]:
```kt
fun PGPPublicKey.getExpirationDate(): Date? =
    validSeconds.takeIf { it != 0L }?.let { Date(creationTime.time + TimeUnit.SECONDS.toMillis(it)) }


```

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/bouncycastle/openpgp/PGPPublicKeyExt.kt [hunk 3]
```diff
@@ -20,3 +27,62 @@ fun PGPPublicKey.armor(
...
+  val isRevoked = hasRevocation()
+  val isExpired = getExpirationDate()?.let { System.currentTimeMillis() > it.time } ?: false
+
+  return ContextCompat.getColorStateList(
+    context, when {
+      isRevoked -> R.color.red
+      isExpired -> R.color.orange
+      else -> R.color.colorPrimary
+    }
+  )
```
Comment (ioanmo226) [inline FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/bouncycastle/openpgp/PGPPublicKeyExt.kt:66]:
```kt
  val colorRes = when {
    hasRevocation() -> R.color.red
    getExpirationDate()?.before(Date()) == true -> R.color.orange
    else -> R.color.colorPrimary
  }
  return ContextCompat.getColorStateList(context, colorRes)

```

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/pgpainless/key/info/KeyRingInfoExt.kt [hunk 4]
```diff
@@ -5,35 +5,145 @@
...
+          0,
+          insetRight,
+          0
+        )
+        setLayerGravity(i, Gravity.CENTER_VERTICAL)
+      }
+    }
+  } else {
+    array.firstOrNull()
+  }
```
Comment (ioanmo226) [inline FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/pgpainless/key/info/KeyRingInfoExt.kt:117]: How about write like below?

```kt
fun KeyRingInfo.generateKeyCapabilitiesDrawable(context: Context, keyId: Long): Drawable? {
  val keyFlags = getKeyFlagsOf(keyId)
  val drawables = listOf(
    KeyFlag.CERTIFY_OTHER to R.drawable.ic_possibility_cert,
    KeyFlag.ENCRYPT_COMMS to R.drawable.ic_possibility_encryption,
    KeyFlag.ENCRYPT_STORAGE to R.drawable.ic_possibility_encryption,
    KeyFlag.SIGN_DATA to R.drawable.ic_possibility_sign,
    KeyFlag.AUTHENTICATION to R.drawable.ic_possibility_auth
  ).mapNotNull { (flag, drawableRes) ->
    if (keyFlags.contains(flag)) ContextCompat.getDrawable(context, drawableRes) else null
  }

  return when {
    drawables.size > 1 -> LayerDrawable(drawables.toTypedArray()).apply {
      for (i in drawables.indices) {
        setLayerInset(i, drawables.take(i).sumOf { it.intrinsicWidth }, 0,
          drawables.takeLast(drawables.size - i - 1).sumOf { it.intrinsicWidth }, 0)
        setLayerGravity(i, Gravity.CENTER_VERTICAL)
      }
    }
    else -> drawables.firstOrNull()
  }
}
```

**Review by ioanmo226 — COMMENTED** [review]
Nice improvement 👍 
Small comments

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/pgpainless/key/info/KeyRingInfoExt.kt (hunk 4 above)
Comment (DenBond7) [reply FlowCrypt/src/main/java/com/flowcrypt/email/extensions/org/pgpainless/key/info/KeyRingInfoExt.kt]: I've applied this code with a few changes

**Review by ioanmo226 — APPROVED** [review]

**Pull request merged**